### PR TITLE
Upgrade the LSP Implementation of the Language Server

### PR DESCRIPTION
**Pull request by camrein**
The PR intends to upgrade the underlying implementation of the language server protocol. With this upgrade, I've also applied some more changes to the language server:

- Due to compatibility issues I had to switch the logging framework from NLog to Serilog.
- The logging configuration is now made through the *DafnyLanguageServer.appsettings.json* file. Furthermore, it's now possible to change settings such as the *--documents:verify* through this file instead of the command-line option. The configuration file is optional and won't lead to startup failures when its missing.
- General code cleanup based on suggestions by the IDE.

### CODE REVIEW

**Comment by RustanLeino** [comment]
The build gives lots of warnings that suggest using `Async` as part of various method names. I suppose that's a worthwhile warning. Can you please make those changes?

Also, we're just on the brink of cutting a Dafny 3.2. This PR seems nice to include in it. Please let me know if you have thoughts about other things that "must" go in now.

**Comment by camrein** [comment]
Since the warnings all affected the names of the test methods, I have opted to suppress the warning in this case. I don't see a benefit in polluting the names of the test methods with the *Async* suffix. However, if you prefer it this way, I can adapt the methods' names accordingly.

Good to hear about Dafny 3.2. I will gladly prepare the VSCode extension for the upcoming release. I don't think I have any changes that are necessary to include in this release.

File: Source/DafnyLanguageServer/Handlers/DafnyDefinitionHandler.cs [hunk 1]
```diff
@@ -29,23 +29,24 @@ public class DafnyDefinitionHandler : DefinitionHandler {
...
       DafnyDocument? document;
       if(!_documents.TryGetDocument(request.TextDocument, out document)) {
-        _logger.LogWarning("location requested for unloaded document {}", request.TextDocument.Uri);
+        _logger.LogWarning("location requested for unloaded document {DocumentUri}", request.TextDocument.Uri);
         return Task.FromResult(new LocationOrLocationLinks());
       }
       ILocalizableSymbol? symbol;
       if(!document.SymbolTable.TryGetSymbolAt(request.Position, out symbol)) {
-        _logger.LogDebug("no symbol was found at {} in {}", request.Position, request.TextDocument);
+        _logger.LogDebug("no symbol was found at {Position} in {Document}", request.Position, request.TextDocument);
```
Comment (keyboardDrummer) [inline Source/DafnyLanguageServer/Handlers/DafnyDefinitionHandler.cs:37]: What's the effect of naming these template variables? This seems to correspond with [the documentation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.loggerextensions.logdebug?view=dotnet-plat-ext-5.0#microsoft-extensions-logging-loggerextensions-logdebug(microsoft-extensions-logging-ilogger-system-s) but I don't understand the effect.

Comment (camrein) [reply Source/DafnyLanguageServer/Handlers/DafnyDefinitionHandler.cs:37]: Personally, I was not too fond of naming the placeholders either, having to name the placeholders. But it appears to me that just using `{}` is an undocumented feature that works with the default logging implementation as well as with NLog. Unfortunately, it broke with the switch to Serilog. I'd have preferred to stick to NLog, but it just didn't want to work with the updated LSP backend.
When using the logger solely with file/console outputs, it hardly has any use. In contrast, if you're logging into a database or in a structured format such as JSON, you may store key/value pairs for these placeholders.

**Review by RustanLeino — APPROVED** [review]

**Pull request merged**